tests/test_numdicts/test_numdicts.py

```diff
@@ -0,0 +1,413 @@
+from itertools import product
+import pyClarion.numdicts as nd
+
+import unittest
+import math
+import itertools
+
+from pyClarion.numdicts.numdicts import NumDict
+
+
+class TestNumdicts(unittest.TestCase):
+    # TODO fix/improve look in the iter tools library like product
+    def sequenceGenerator(self, a, b):
+        r = 4  # represents how many divisions
+        a = a*r
+        b = b*r
+        for i, j in itertools.product(tuple(range(a, b)), tuple(range(a, b))):
+            yield i/r, j/r
+
+    def test_addition_basic_functionality(self):
+        for i, j in self.sequenceGenerator(-10, 10):
+            tape = nd.GradientTape()
+            with tape:
+                # testing basic functionality for defaults
```

Comment:
This approach to testing defaults does not account for situations like the following.

This case is about defaults being added to keys.
```python
a = NumDict({1: 2.7, 2: 3.4}, default=.7)
b = NumDict({1: 3.9, 3: 4.6}, default=.0)
assertAlmostEqual(a + b,  NumDict({1: 6.6, 2: 3.4 3: 5.3}, default=.7))
```

This case is when default is not defined, I would assume this raises a `KeyError` right now because lookup fails on `3`. At any rate, it should raise *some* error, probably derivative from `KeyError`.
```python
a = NumDict({1: 2.7, 2: 3.4})
b = NumDict({1: 3.9, 3: 4.6}, default=.0)
with assertRaises(KeyError):
    a + b
```

So, in general there seem to be four basic things to check, as listed below.
1. Does the op apply to defaults correctly?
2. Does the op apply to keys correctly?
3. Does the op apply to one key and one default correctly?
4. Does the op fail appropriately when a key is missing and no default is given?

These checks are probably best treated as separate test methods.

The last check also has implications for the implementation, it may be worth having an error that is more specific than `KeyError`. I generally avoid writing custom exceptions, but this may be a case where, if nothing else, a special error message may be appropriate.